### ISSUE TITLE
Add support for when clause in commands

### ISSUE THREAD

**Issue by avishnyak** [VSCodeVim/Vim#2030]
* Click *thumbs-up* 👍 on this issue if you want it!
* Click *confused* 😕 on this issue if not having it makes VSCodeVim unusable. 

The VSCodeVim team prioritizes issues based on reaction count. 

--------

**FEATURE REQUEST**:

Support "[when](https://code.visualstudio.com/docs/getstarted/keybindings#_when-clause-contexts)" in custom binding commands.  This will allow one to create toggle commands that do different things based on the state of the editor.

For example:
```js
"vim.otherModesKeyBindingsNonRecursive": [
        {
            "before": ["-"],
            "commands": [
                {
                    "command": "workbench.files.action.focusFilesExplorer",
                    "when": "editorFocus",
                    "args": []
                },
                {
                    "command": "workbench.action.toggleSidebarVisibility",
                    "when": "explorerViewletFocus",
                    "args": []
                }
            ]
        }
    ]
```

Pressing ```-``` while focused on an editor, shows and sets focus to the file explorer.  Pressing ```-``` in the file explorer, closes the explorer sidebar and sets you back into the editor.

**Comment by xconverge** [comment]
I can't figure out a good way to query when, vscode does not expose this information to the extension

**Comment by nkabbara** [comment]
@avishnyak I was actually also working on mapping '-' to mimic tpope's vim-vinegar. Any luck with this?

**Comment by avishnyak** [comment]
No. The feature had not been implemented. I have to use '-' to focus the file tree and different shortcuts to get back to the buffer.

**Comment by sim590** [comment]
Isn't this a duplicate of #1372 ?
